### PR TITLE
Reqs finance check + create order

### PR DESCRIPTION
**Pull request by iaincrawf**
I also need to copy a couple of wrapper functions into prod db before putting this live

- Big thing missing is emails to suppliers with the purchase order and blurb attached, will try do that next
- need to update internal comments on mini order when turning req to order, might add it to my wrapper or if it's mapped already maybe use that

### CODE REVIEW

File: src/Domain.LinnApps/PurchaseOrderReqs/PurchaseOrderReqService.cs [hunk 1]
```diff
@@ -110,6 +124,86 @@ public PurchaseOrderReq Create(PurchaseOrderReq entity, IEnumerable<string> priv
...
+            }
+
+            if (!entity.AuthorisedById.HasValue)
+            {
+                throw new UnauthorisedActionException("Cannot create order from a req that has not been authorised");
+            }
+
+            if (!entity.TotalReqPrice.HasValue)
+            {
+                throw new UnauthorisedActionException("Cannot create order from a req without value for price");
```
Comment (lewisrenfrew) [inline src/Domain.LinnApps/PurchaseOrderReqs/PurchaseOrderReqService.cs:142]: would be tempted to use a different exception type here

File: src/Facade/ResourceBuilders/PurchaseOrderReqResourceBuilder.cs [hunk 2]
```diff
@@ -157,7 +158,12 @@ private IEnumerable<LinkResource> BuildLinks(PurchaseOrderReq model, IEnumerable
 
                 if (this.authService.HasPermissionFor(AuthorisedAction.PurchaseOrderCreate, privileges))
                 {
-                    yield return new LinkResource { Rel = "create-purchase-order", Href = this.GetLocation(model) };
+                    yield return new LinkResource { Rel = "turn-req-into-purchase-order", Href = $"{this.GetLocation(model)}/turn-into-order" };
```
Comment (lewisrenfrew) [inline src/Facade/ResourceBuilders/PurchaseOrderReqResourceBuilder.cs:161]: possibly untested links building logic?

File: src/Facade/Services/PurchaseOrderReqFacadeService.cs [hunk 3]
```diff
@@ -67,19 +67,55 @@ public class PurchaseOrderReqFacadeService : FacadeFilterResourceService<Purchas
...
+            var entity = this.repository.FindById(reqNumber);
+            if (entity == null)
+            {
+                return new NotFoundResult<ProcessResultResource>();
+            }
+
+           var result = this.domainService.CheckIfSigningLimitCanAuthorisePurchaseOrder(entity, currentUserNumber);
+
+           return new SuccessResult<ProcessResultResource>(new ProcessResultResource(result.Success, result.Message));
+
```
Comment (lewisrenfrew) [inline src/Facade/Services/PurchaseOrderReqFacadeService.cs:85]: blank line

File: src/Proxy/PurchaseOrderAutoOrder.cs [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+                    Value = authAllowed ? 1 : 0
+                };
+
+                var result = new OracleParameter(null, OracleDbType.Int32)
+                {
+                    Direction = ParameterDirection.ReturnValue
+                };
+
+                //FUNCTION Create_Auto_Order_Wrapper(p_part varchar2, p_supplier number, p_qty number, p_date date,
+                //                                   p_order_number out number, p_our_price number default null,
```
Comment (lewisrenfrew) [inline src/Proxy/PurchaseOrderAutoOrder.cs:162]: delete?

File: src/Proxy/PurchaseOrdersPack.cs [hunk 5]
```diff
@@ -48,5 +49,76 @@ public bool OrderIsCompleteSql(int orderNumber, int lineNumber)
...
+                                 };
+
+                var pdocumenttype = new OracleParameter("p_document_type", OracleDbType.Varchar2)
+                                {
+                                    Direction = ParameterDirection.Input,
+                                    Size = 50,
+                                    Value = documentType
+                };
+
+
```
Comment (lewisrenfrew) [inline src/Proxy/PurchaseOrdersPack.cs:103]: blankity blank

File: src/Proxy/PurchaseOrdersPack.cs [hunk 6]
```diff
@@ -48,5 +49,76 @@ public bool OrderIsCompleteSql(int orderNumber, int lineNumber)
...
+                };
+
+                var ptotal = new OracleParameter("p_total", OracleDbType.Decimal)
+                                {
+                                    Direction = ParameterDirection.Input,
+                                    Size = 50,
+                                    Value = totalValueBase
+                                };
+
+
```
Comment (lewisrenfrew) [inline src/Proxy/PurchaseOrdersPack.cs:88]: blankity blank

File: tests/Unit/Domain.LinnApps.Tests/PurchaseOrderReqServiceTests/ContextBase.cs [hunk 7]
```diff
@@ -30,6 +29,13 @@ public class ContextBase
 
         protected IPurchaseOrderReqService Sut { get; private set; }
 
+        protected IPurchaseOrderAutoOrderPack MockPurchaseOrderAutoOrderPack { get; private set; }
+
+        protected ICurrencyPack MockCurrencyPack { get; private set; }
+        protected IPurchaseOrdersPack MockPurchaseOrdersPack { get; private set; }
+
+
+
```
Comment (lewisrenfrew) [inline tests/Unit/Domain.LinnApps.Tests/PurchaseOrderReqServiceTests/ContextBase.cs:38]: blanks

File: tests/Unit/Domain.LinnApps.Tests/PurchaseOrderReqServiceTests/WhenCheckingCanAuthPlOrder.cs [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+            this.MockPurchaseOrdersPack.OrderCanBeAuthorisedBy(
+                null,
+                null,
+                this.authoriserUserNumber,
+                145m,
+                this.entity.PartNumber,
+                "PO").Returns(false);
+
+
```
Comment (lewisrenfrew) [inline tests/Unit/Domain.LinnApps.Tests/PurchaseOrderReqServiceTests/WhenCheckingCanAuthPlOrder.cs:78]: blank

File: tests/Unit/Domain.LinnApps.Tests/PurchaseOrderReqServiceTests/WhenFinanceChecking.cs [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+                              };
+
+            this.MockReqsStateChangeRepository.FindBy(Arg.Any<Expression<Func<PurchaseOrderReqStateChange, bool>>>())
+                .Returns(
+                    new PurchaseOrderReqStateChange
+                        {
+                            FromState = this.fromState, ToState = this.toState, UserAllowed = "Y"
+                        });
+
+
```
Comment (lewisrenfrew) [inline tests/Unit/Domain.LinnApps.Tests/PurchaseOrderReqServiceTests/WhenFinanceChecking.cs:75]: blank

File: tests/Unit/Domain.LinnApps.Tests/PurchaseOrderReqServiceTests/WhenFinanceCheckingAndNotAllowed.cs [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+                                  SecondAuthById = null,
+                                  FinanceCheckById = null,
+                                  TurnedIntoOrderById = null,
+                                  NominalCode = "00001234",
+                                  RemarksForOrder = "needed asap",
+                                  InternalNotes = "pls approv",
+                                  DepartmentCode = "00002345"
+                              };
+
+
```
Comment (lewisrenfrew) [inline tests/Unit/Domain.LinnApps.Tests/PurchaseOrderReqServiceTests/WhenFinanceCheckingAndNotAllowed.cs:71]: blank

File: tests/Unit/Domain.LinnApps.Tests/PurchaseOrderReqServiceTests/WhenFinanceCheckingAndNotAllowed.cs [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+            this.MockAuthService.HasPermissionFor(
+                AuthorisedAction.PurchaseOrderReqFinanceCheck,
+                Arg.Any<List<string>>()).Returns(false);
+
+            this.action = () => this.Sut.FinanceApprove(this.entity, new List<string>(), this.authoriserUserNumber);
+        }
+
+        [Test]
+        public void ShouldThrowIllegalStateException()
```
Comment (lewisrenfrew) [inline tests/Unit/Domain.LinnApps.Tests/PurchaseOrderReqServiceTests/WhenFinanceCheckingAndNotAllowed.cs:80]: test name?

File: src/Domain.LinnApps/PurchaseOrderReqs/PurchaseOrderReqService.cs [hunk 12]
```diff
@@ -80,11 +92,13 @@ public void Authorise(PurchaseOrderReq entity, IEnumerable<string> privileges, i
             {
                 entity.State = allowedToAuthoriseResult.NewState;
                 entity.AuthorisedById = currentUserId;
+                entity.AuthorisedBy = this.employeeRepository.FindById(currentUserId);
```
Comment (lewisrenfrew) [inline src/Domain.LinnApps/PurchaseOrderReqs/PurchaseOrderReqService.cs:95]: from a quick look a lot of these DomainService if (..) statements don't have corresponding unit test for WhenThatScenario.. green ncrunch dots are your friend

File: src/Domain.LinnApps/PurchaseOrderReqs/PurchaseOrderReqService.cs [hunk 13]
```diff
@@ -140,7 +236,13 @@ public ProcessResult SendAuthorisationRequestEmail(int currentUser, int toEmp, P
                 this.emailService.SendEmail(
                     to.PhoneListEntry.EmailAddress.Trim(),
                     to.FullName,
-                    new List<Dictionary<string, string>> { new Dictionary<string, string>() { { "name", from.FullName }, { "address", from.PhoneListEntry.EmailAddress.Trim() } } },
+                    new List<Dictionary<string, string>>
+                        {
+                            new Dictionary<string, string>
```
Comment (lewisrenfrew) [inline src/Domain.LinnApps/PurchaseOrderReqs/PurchaseOrderReqService.cs:241]: again, unit tests to check the correct emails are sent?

**Review by lewisrenfrew — APPROVED** [review]

File: src/Domain.LinnApps/PurchaseOrderReqs/PurchaseOrderReqService.cs (hunk 1 above)
Comment (iaincrawf) [reply src/Domain.LinnApps/PurchaseOrderReqs/PurchaseOrderReqService.cs:142]: Yeah that's probably fair, think a DomainException or a custom "no value" sort of one? A generic-y RequiredButNoValueException or a slightly improved name could be useful elsewhere I'm sure

Comment (iaincrawf) [reply src/Domain.LinnApps/PurchaseOrderReqs/PurchaseOrderReqService.cs:142]: might use ArgumentException

**Pull request merged**